WcaOnRails/app/controllers/api/v0/competitions_controller.rb

```diff
@@ -1,11 +1,62 @@
 # frozen_string_literal: true
 class Api::V0::CompetitionsController < Api::V0::ApiController
-  def show
-    competition = Competition.where(id: params[:id], showAtAll: true).first
+  def competition_from_params
+    id = params[:competition_id] || params[:id]
+    competition = Competition.visible.where(id: id).first
     unless competition
-      render json: { error: "Competition with id #{params[:id]} not found" }, status: 404
-      return
+      render json: { error: "Competition with id #{id} not found" }, status: 404
+      return nil
     end
+    competition
+  end
+
+  def index
+    competitions = Competition.search(params[:q], params: params, current_resource_owner: current_resource_owner)
+    competitions = competitions.includes(:delegates).includes(:organizers)
+
+    paginate json: competitions
+  end
+
+  def show
+    competition = competition_from_params
+    return unless competition
+
     render json: competition
   end
+
+  def can_manage_competition?(competition)
+    unless current_resource_owner
+      render json: { error: "Not authenticated" }, status: 401
+      return false
+    end
+
+    unless doorkeeper_token.scopes.exists?("manage_competitions") && current_resource_owner.can_manage_competition?(competition)
+      render json: { error: "Not authorized to manage competition" }, status: 403
+      return false
+    end
+
+    true
+  end
+
+  def big_json
+    competition = competition_from_params
+    return unless competition
+    return unless can_manage_competition?(competition)
+
+    render json: competition.big_json
+  end
+
+  def update_big_json
+    competition = competition_from_params
+    return unless competition
+    return unless can_manage_competition?(competition)
+
+    new_big_json = params.require(:big_json).permit!.to_h
+    competition.big_json = new_big_json
+    if competition.save
```

Comment:
Hmm... we have talked about never calling `.save` and ignoring its return value. It's perfectly legitimate to call plain `.save` and render something special if it returns `false` (which is analogous to what I'm doing here by rendering a 400 with the errors when `.save` fails). AFAIK, that's the only way to handle validation errors with Rails forms.